internal/impl/redpanda/functions.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redpanda
+
+import (
+	"context"
+
+	"github.com/tetratelabs/wazero/api"
+)
+
+const (
+	noActiveTransform = int32(-1)
+	invalidBuffer     = int32(-2)
+)
+
+var transformHostFunctions = map[string]func(r *dataTransformEngine) interface{}{}
+
+func registerModuleRunnerFunction(name string, ctor func(r *dataTransformEngine) interface{}) struct{} {
+	transformHostFunctions[name] = ctor
+	return struct{}{}
+}
+
+var _ = registerModuleRunnerFunction("check_abi_version_1", func(r *dataTransformEngine) interface{} {
+	return func(ctx context.Context, m api.Module) {
+		// Placeholder for ABI compatibility check
+	}
+})
+var _ = registerModuleRunnerFunction("check_abi_version_2", func(r *dataTransformEngine) interface{} {
+	return func(ctx context.Context, m api.Module) {
+		// Placeholder for ABI compatibility check
+	}
+})
+
+var _ = registerModuleRunnerFunction("read_batch_header", func(r *dataTransformEngine) interface{} {
+	return func(
+		ctx context.Context,
+		m api.Module,
+		baseOffset,
+		recordCount,
+		partitionLeaderEpoch,
+		attributes,
+		lastOffsetDelta,
+		baseTimestamp,
+		maxTimestamp,
+		producerId,
+		producerEpoch,
+		baseSequence uint32) int32 {
+		// Notify the host we're done processing a batch.
+		r.hostChan <- nil
+		// Wait for new batch to be submitted for processing.
+		select {
+		case _, ok := <-r.guestChan:
+			if !ok {
+				return noActiveTransform
+			}
+		case <-ctx.Done():
+			return noActiveTransform
+		}
+		if !m.Memory().WriteUint32Le(recordCount, uint32(len(r.inputBatch))) {
```

Comment:
Could `m.Memory()` ever be nil? (same in the rest of the functions)

Reply:
Ah good catch - I will validate that in newDataTransformProcessor :)